src/toast/pixels.py

```diff
@@ -1162,6 +1170,529 @@ def broadcast_map(self, fdata, comm_bytes=10000000):
+                                local_submap = dist.global_submap_to_local[submap]
+                                sendbuffer[offset] = self.data[local_submap].T
+                                offset += 1
+
+                    if rank == 0:
+                        # Root process receives the submaps from other processes and writes
+                        # them to file
+                        with h5py.File(path, "w") as f:
+                            dset = f.create_dataset(
+                                "map",
+                                (self.n_value, dist.n_pix),
+                                chunks=(self.n_value, dist.n_pix_submap),
+                                dtype=dtype,
+                            )
+                            for rank_send in range(nproc):
+                                # submaps = np.argwhere(dist.submap_owners == rank_send).ravel()
+                                submaps = np.arange(dist.n_submap)[
+                                    allowners == rank_send
+                                ]
+                                n_receive = len(submaps)
+                                if n_receive > 0:
+                                    if rank_send == rank:
+                                        recvbuffer = sendbuffer
+                                    else:
+                                        recvbuffer = np.empty(
+                                            [
+                                                n_receive,
+                                                self.n_value,
+                                                dist.n_pix_submap,
+                                            ],
+                                            dtype=dtype,
+                                        )
+                                        dist.comm.Recv(
+                                            recvbuffer, source=rank_send, tag=rank_send
+                                        )
+                                for i, submap in enumerate(submaps):
+                                    # Accommodate submap sizes that do not fit the map cleanly
+                                    first = submap * dist.n_pix_submap
+                                    last = min(first + dist.n_pix_submap, dist.n_pix)
+                                    dset[:, first:last] = recvbuffer[
+                                        i, :, 0 : last - first
+                                    ]
+
+                            for key, value in header.items():
+                                dset.attrs[key] = value
+                    else:
+                        # All others wait for their turn to send
+                        if sendbuffer is not None:
+                            dist.comm.Send(sendbuffer, dest=0, tag=rank)
+            else:
+                msg = f"Could not determine file type for '{path}'"
+                raise RuntimeError(msg)
+
+    @function_timer
+    def read(
```

Comment:
This is a very long method and would benefit from being refactored into a thin wrapper around HEALPix and WCS methods. That should not hold off on merging, just an observation.

Reply:
I agree- we could have internal methods for the different implementations.

Reply:
I have broken up the read / write methods to be thin wrappers around the specific implementations of pixelization and file formats.